### PR TITLE
add core functions make_timestamp_ns(nanos) and epoch_ns(timestamp_ts)

### PR DESCRIPTION
**Pull request by andreimatei**
See individual commits.

### CODE REVIEW

**Comment by andreimatei** [comment]
cc @hawkfish

I have a feeling that the new `make_timestamp_ns` function should be listed in `extension_entries.hpp`, but I have failed to generate that file correctly. I've tried invoking it in different ways, but no matter what I do, the `generate_extensions_function.py` script deletes all `core_functions` entries for me. Perhaps I could get help with the right entry. Thanks!

**Review by hawkfish — CHANGES_REQUESTED** [review]
Very minor stuff.

File: src/include/duckdb/common/types/timestamp.hpp [hunk 1]
```diff
@@ -24,6 +24,8 @@ struct dtime_tz_t; // NOLINT
 
 //! Type used to represent a TIMESTAMP. timestamp_t holds the microseconds since 1970-01-01.
 struct timestamp_t { // NOLINT
+	// NOTE: The unit of value is microseconds for timestamp_t, but it can de
```
[the remark this answers is not in the document]
Comment (andreimatei) [reply src/include/duckdb/common/types/timestamp.hpp]: done

File: test/sql/function/timestamp/test_date_part.test [hunk 2]
```diff
@@ -961,3 +961,9 @@ statement error
 WITH parts(p) AS (VALUES (['year', 'month', 'day']), (['hour', 'minute', 'microsecond']))
 SELECT DATE_PART(p, ts) FROM parts, timestamps;
 ----
+
+# epoch_ns has nanosecond precision on TIMESTAMP_TS
```
[the remark this answers is not in the document]
Comment (andreimatei) [reply test/sql/function/timestamp/test_date_part.test]: done

File: extension/core_functions/scalar/date/make_date.cpp [hunk 3]
```diff
@@ -98,7 +98,7 @@ struct MakeTimestampOperator {
 
 	template <typename T, typename RESULT_TYPE>
 	static RESULT_TYPE Operation(T micros) {
-		return timestamp_t(micros);
+		return RESULT_TYPE(micros);
```
[the remark this answers is not in the document]
Comment (andreimatei) [reply extension/core_functions/scalar/date/make_date.cpp]: separated change to its own commit and done

**Review by taniabogatsch — APPROVED** [review]

File: src/include/duckdb/common/types/timestamp.hpp [hunk 4]
```diff
@@ -189,6 +191,7 @@ class Timestamp {
 	DUCKDB_API static int64_t GetEpochMicroSeconds(timestamp_t timestamp);
 	//! Convert a timestamp to epoch (in nanoseconds)
 	DUCKDB_API static int64_t GetEpochNanoSeconds(timestamp_t timestamp);
+	DUCKDB_API static int64_t GetEpochNanoSeconds(timestamp_ns_t timestamp);
```
Comment (taniabogatsch) [inline src/include/duckdb/common/types/timestamp.hpp:194]: This is the same as directly accessing `value`, no? I'm unsure if we need it, but I'm also okay with adding it.

File: test/sql/function/timestamp/make_date.test [hunk 5]
```diff
@@ -217,6 +217,12 @@ SELECT make_timestamp(0), make_timestamp(1684509234845000);
 ----
 1970-01-01 00:00:00	2023-05-19 15:13:54.845
 
+# From nanoseconds
+query II
+SELECT make_timestamp_ns(0), make_timestamp_ns(1684509234845000123);
+----
+1970-01-01 00:00:00	2023-05-19 15:13:54.845000123
+
```
Comment (taniabogatsch) [inline test/sql/function/timestamp/make_date.test:225]: Should we add tests for `min` and `max` values?

**Review by taniabogatsch — COMMENTED** [review]
@hawkfish, indeed, there have been a few changes to the `s, ms, ns` types. See https://github.com/duckdb/duckdb/pull/14818.

File: src/include/duckdb/common/types/timestamp.hpp (hunk 4 above)
Comment (andreimatei) [reply src/include/duckdb/common/types/timestamp.hpp:194]: > This is the same as directly accessing value, no?

Yes. But so is [`GetEpochMicroSeconds(timestamp_t timestamp)`](https://github.com/andreimatei/duckdb/blob/da42e2d9fcebbc302547dda516b546bda6577870/src/common/types/timestamp.cpp#L430-L432) -- and that function exists.
So, if it's up to me, I'd leave it. But I don't know this code at all, so...

File: test/sql/function/timestamp/make_date.test (hunk 5 above)
Comment (andreimatei) [reply test/sql/function/timestamp/make_date.test:225]: Do you mean `make_timestamp_ns(-9,223,372,036,854,775,808)` and `make_timestamp_ns(9,223,372,036,854,775,807)` ?

**Review by hawkfish — APPROVED** [review]

File: test/sql/function/timestamp/make_date.test (hunk 5 above)
Comment (andreimatei) [reply test/sql/function/timestamp/make_date.test:225]: I've added a test for the maximum values. 

For negative values, I've ran into some problems:
`make_timestamp_ns(-1)` triggers an assertion:
```
select make_timestamp_ns(-1);
INTERNAL Error:
Information loss on integer cast: value -952 outside of target range [-128, 127]
```
I couldn't immediately figure out why, or how to get a stack trace for the respective exception.

`select make_timestamp(-9223372036854775807)` correctly says `-infinity`.

`make_timestamp_ns(-9223372036854775806)` says:
```
select make_timestamp_ns(-9223372036854775806);
Conversion Error:
Date out of range in timestamp_ns conversion
```
This is similar to `make_timestamp_ns(-9223372036854775806)`. I don't know if the behavior is good; I couldn't find a lower limit for `make_timestamp()` documented anywhere.

Would it be acceptable to reject any negative argument to `make_timestamp_ns()` ? I'll kindly take advice for how to proceed / if you could look into the assertion error if rejecting negative input is not acceptable, I'd appreciate it.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**